### ISSUE TITLE
PVR Recordings are reported to be seekable.  

### ISSUE THREAD

**Issue by emveepee** [xbmc/xbmc#18368]
## Bug report
Core doesn't understand that a recording may not be seekable.   On some recordings core my closes a recording,  but continue to perform actions before closing the recording a second time.
### Describe the bug
Here is a clear and concise description of what the problem is:

After a recording is closed core issues a CloseRecordedStream() to the PVR addon but could follow up with GetStreamTimes(), PauseStream(), IsRealTimeStream()  and SeekRecordedStream() on the closed file in my observations.

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

After a CloseRecordedStream()  code should not be trying to access the file via the addon

## Actual Behavior

Best to see the logs

```
2020-09-03 10:33:14.607 T:10296   DEBUG <general>: CInputStreamPVRRecording - CInputStreamPVRRecording::ClosePVRStream - closed recording stream pvr://recordings/tv/active/Just for Laughs: Gags/Just%20for%20Laughs%3a%20Gags, TV%20(MUCH), 20200522_161449, 22764.pvr
....
2020-09-03 10:33:14.607 T:10296   ERROR <general>: GetStreamTimes: Add-on 'NextPVR:172.16.3.105' returned an error: invalid parameters for this method
...
2020-09-03 10:33:14.608 T:10296   DEBUG <general>: demuxer seek to: 69456.485000
2020-09-03 10:33:14.608 T:10296   DEBUG <general>: ffmpeg[0x2a97e048ed0X]: [mpegts] read_seek: 0 1049040000
2020-09-03 10:33:14.608 T:10296   DEBUG <general>: ffmpeg[0x2a97e048ed0X]: [mpegts] gen_seek: 0 1049040000
2020-09-03 10:33:14.608 T:10296   DEBUG <general>: AddOnLog: pvr.nextpvr: Seek: timeshift::RecordingBuffer::Seek:37  0  -1 -1 -1
...
2020-09-03 10:33:14.619 T:10296   ERROR <general>: GetStreamTimes: Add-on 'NextPVR:172.16.3.105' returned an error: invalid parameters for this method
...
2020-09-03 10:33:14.641 T:10296   DEBUG <general>: CInputStreamPVRRecording - CInputStreamPVRRecording::ClosePVRStream - closed recording stream pvr://recordings/tv/active/Just for Laughs: Gags/Just%20for%20Laughs%3a%20Gags, TV%20(MUCH), 20200522_161449, 22764.pvr
2020-09-03 10:33:14.641 T:10296   DEBUG <general>: Thread VideoPlayer 10296 terminating
2020-09-03 10:33:14.645 T:15188    INFO <general>: Deleting settings information for files pvr://recordings/tv/active/Just for Laughs: Gags/Just%20for%20Laughs%3a%20Gags, TV%20(MUCH), 20200522_161449, 22764.pvr
2020-09-03 10:33:14.646 T:15256   DEBUG <general>: CSaveFileState::DoWork - Saving file state for video item pvr://recordings/tv/active/Just for Laughs: Gags/Just%20for%20Laughs%3a%20Gags, TV%20(MUCH), 20200522_161449, 22764.pvr
```

## Possible Fix

I believe it is file specific since I don't see this on all recordings. 

### To Reproduce

1. Play certain recordings from PVR
2. Navigate or pauses
3. If actions fails two stops can be issued

### Debuglog

The debuglog can be found here:

https://paste.kodi.tv/usipenicil.kodi

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

A user provided this file https://www.transfernow.net/kjM8js082020  

It is a "bad" file in that in the pre-padding area the station was off air and ffmpeg is not handling it well.   There are issues playing the file in the file library.  

This issue is not directly related to the file since core should consider the recording closed if a CloseRecordedStream() is issued to the addon

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] Linux
 - [ ] OSX
 - [ ] Raspberry-Pi
 - [x ] Windows
 - [ ] Windows UWP

 - Operating system version/name:
 - Kodi version:
Current nightly plus self builds.

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by phunkyfish** [comment]
Is the crash in the addon or in core?

**Comment by emveepee** [comment]
> 
> 
> Is the crash in the addon or in core?

For this issue am using a custom addon that is not crashing .  The crash you are referring too is likely the linked issue I created for the addon.

**Comment by phunkyfish** [comment]
Ah, ok. I’m not sure if core calls close more once it should be an issue. It’s more important it calls it at least once I would think.

**Comment by emveepee** [comment]
The issue is calling GetStreamTimes() seeks etc after the file is closed.

**Comment by phunkyfish** [comment]
Yes, but each of those calls should be guarded in the addon. I.e. if stream is valid: do stuff.

**Comment by emveepee** [comment]
Yes agreed that is why I created two issues.  There isn't even a good PVR_ERROR code for bad API calls and SeekRecordededStream() doesn't define this error condition and will treat -1  like PVR_ERROR_NOT_IMPLEMENTED

**Comment by FernetMenta** [comment]
PVR recordings are not supposed to use CFileCache.

**Comment by emveepee** [comment]
> 
> 
> PVR recordings are not supposed to use CFileCache.

Thanks very much I did not know this and removed the OPEN_CACHED flag  I was using for concluded files.  Unfortunately I still get two closes with a seek and GetStreamTimes() being called in between them

```
2020-09-06 12:25:12.877 T:17948   DEBUG <general>: AddOnLog: pvr.nextpvr: Buffer::Open() called! [ /home/martin/recordings/Just for Laughs Gags/Just for Laughs Gags_20200522_12001230.ts ]
2020-09-06 12:25:12.909 T:17948   DEBUG <general>: CInputStreamPVRRecording - CInputStreamPVRRecording::OpenPVRStream - opened recording stream pvr://recordings/tv/active/Just for Laughs: Gags/Just%20for%20Laughs%3a%20Gags, TV%20(MUCH), 20200522_161449, 22764.pvr
2020-09-06 12:25:12.909 T:17948    INFO <general>: Creating Demuxer
2020-09-06 12:25:12.911 T:17948   DEBUG <general>: CDVDDemuxFFmpeg::Open - probing detected format [mpegts]
...
2020-09-06 12:25:23.688 T:17948   DEBUG <general>: AddOnLog: pvr.nextpvr: timeshift::Buffer::CloseHandle:64:
2020-09-06 12:25:23.688 T:17948   ERROR <general>: AddOnLog: pvr.nextpvr: Duration set to 0 XXXXX
2020-09-06 12:25:23.689 T:17948   DEBUG <general>: CInputStreamPVRRecording - CInputStreamPVRRecording::ClosePVRStream - closed recording stream pvr://recordings/tv/active/Just for Laughs: Gags/Just%20for%20Laughs%3a%20Gags, TV%20(MUCH), 20200522_161449, 22764.pvr
2020-09-06 12:25:23.689 T:17948   DEBUG <general>: CDVDDemuxFFmpeg::SeekTime - seek ended up on time -2147483648
2020-09-06 12:25:23.689 T:17948   DEBUG <general>: VideoPlayer: seek failed or hit end of stream
2020-09-06 12:25:23.689 T:17948   DEBUG <general>: CVideoPlayer::FlushBuffers - flushing buffers
2020-09-06 12:25:23.689 T:17948   ERROR <general>: AddOnLog: pvr.nextpvr: Unknown streaming state 0 0 1 -858993460
2020-09-06 12:25:23.689 T:17948   ERROR <general>: GetStreamTimes: Add-on 'NextPVR:172.16.3.105' returned an error: unknown error
2020-09-06 12:25:23.689 T:17948   ERROR <general>: AddOnLog: pvr.nextpvr: Unknown streaming state 0 0 1 -858993460
2020-09-06 12:25:23.690 T:17948   ERROR <general>: AddOnLog: pvr.nextpvr: Unknown streaming state 0 0 1 -858993460
2020-09-06 12:25:23.690 T:17948   DEBUG <general>: demuxer seek to: 22213.252600
2020-09-06 12:25:23.690 T:17948   ERROR <general>: AddOnLog: pvr.nextpvr: Unknown streaming state 0 0 1 -858993460
2020-09-06 12:25:23.690 T:17948   ERROR <general>: AddOnLog: pvr.nextpvr: Duration set to 0 XXXXX
2020-09-06 12:25:23.691 T:17948   DEBUG <general>: CInputStreamPVRRecording - CInputStreamPVRRecording::ClosePVRStream - closed recording stream pvr://recordings/tv/active/Just for Laughs: Gags/Just%20for%20Laughs%3a%20Gags, TV%20(MUCH), 20200522_161449, 22764.pvr
2020-09-06 12:25:23.691 T:17948   DEBUG <general>: CDVDDemuxFFmpeg::SeekTime - seek ended up on time -2147483648
2020-09-06 12:25:23.691 T:17948   DEBUG <general>: VideoPlayer: seek failed or hit end of stream
```

**Comment by djp952** [comment]
@emveepee, may I suggest that a documentation change might be the way to go here to get this closed out? A note in the interface declaration that these functions may be called at any time whether or not a stream is actually open might do the trick, like the note that was added for OpenLiveStream and OpenRecordedStream that there is an implicit call to CloseLiveStream / CloseRecordedStream that will always occur beforehand.

In this case I'm inclined to agree with @phunkyfish that the addon should ultimately be held responsible for protecting itself and not crashing, but documenting the behavior for future developers can't hurt.  You never know when a change to one part of Kodi might inadvertently end up calling an addon callback at a bad time.  GUI and VideoPlayer enhancements seem to be pretty good at doing that :)

I do think that a more specific error code than NOT_IMPLEMENTED (which is what I happen to use in these cases as well) would be good, perhaps something like INVALID_STATE?  Having a new code to use could be caught during testing of new features and may alter the changeset author that there is a potential problem with the existing addon(s) ...

**Comment by emveepee** [comment]
@djp952 I do agree about the responsibility for preventing the crash and have already raised the issue and fixed it in pvr.nextpvr Still the core shouldn't be reading/seeking etc on a on file that it manages and has already closed .  However in recent nightlies the file I had to demonstrate the issue no longer crashes Kodi so this can probably be closed.

As for documentation that does make sense https://codedocs.xyz/AlwinEsch/kodi/group__cpp__kodi__addon__pvr___streams.html#details could probably have it near the warning  error (which I don't understand BTW)  

Ideally all these calls should use a client supplied file handle for better error checking and for future possibilities like PIP.  Some of the calls I was getting didn't even return PVR_ERROR so the fix is still a hack.

**Comment by emveepee** [comment]
@ksooo  @phunkyfish  I am going to rename this since the root issues appear to be that PVR assumes all recordings are seekable https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/DVDInputStreams/InputStreamPVRRecording.cpp#L96  and it also doesn't handle seek errors.

I can adjust GetStreamTimes() to help when the inputstream doesn't report size, but sometimes ffmpeg will return seek errors which was the case on my sample file and I just have to wait for core to stop trying and handle the errors

**Comment by phunkyfish** [comment]
If you open this file directly in kodi would kodi also assume it is seekable? My guess is that it would but let's verify that.

I don't think that the assumption that recordings are seekable is incorrect.

**Comment by emveepee** [comment]
The core logic includes a Seek(0, SEEK_POSSIBLE) whence which PVR doesn't use and videoplayer will not seek  on stream that ffmpeg marks unseekable by file content.   What is seekable using the file system may not be the same as what is seekable via streaming from a library playlist.  I didn't test audio because I know that Kodi will not try and seek on an internet radio stream.  

This issue became clearer in a recent find that some proxy servers were stripping content-length from recordings (possibly for their cache?) and Kodi was marking these recordings not seekable.

A possible recording scenario for me which we discussed would be to pass an m3u8 file of a recording to inputstream or inputstream.ffmpegdirect and so these aren't necessarily seekable by inputstream.   If VOD was filtered as a recording and not a channel (which to me is quite appropriate) these also might not be seekable at source.

The future PVR API should as a minimum change CanSeekStream()  to something like CanSeekChannelStream() to make it clearer what function it overrides.     CanSeekRecordingStream() might be considered as a valid option but allowing CanSeekStream() to be dual function (like pvr.nextpvr was incorrectly assuming) could easily break existing addons based on a my quick sampling of source.
